### PR TITLE
Reduce allocations and cpu load on AWSSDKUtils.EncodeTraceIdHeaderValue

### PR DESCRIPTION
**Pull request by JanHyka**
## Description
1. Change EncodeTradeIdHeaderValue to use rented buffer for intermediate encoded utf8 value.
2. Use precalculated bool array of whitelisted characters to avoid O(m x n) complexity.
3. Avoid using String.Format to convert byte to hex (way too much overhead plus we already have means for fast allocation-less conversion).
4. Use ValueStringBuilder
5. Fix expected and evaluated values position in assert to avoid confusing error message.

| Method                       | Scenario   | Mean        | Ratio | Gen0   | Allocated | Alloc Ratio |
|----------------------------- |----------- |------------:|------:|-------:|----------:|------------:|
| **EncodeTraceIdHeaderValue**     | **clean**      |   **170.03 ns** |  **1.00** | **0.0343** |     **648 B** |        **1.00** |
| EncodeTraceIdHeaderValue_new | clean      |    81.17 ns |  0.48 | 0.0093 |     176 B |        0.27 |
|                              |            |             |       |        |           |             |
| **EncodeTraceIdHeaderValue**     | **typical**    |   **289.26 ns** |  **1.00** | **0.0529** |    **1000 B** |        **1.00** |
| EncodeTraceIdHeaderValue_new | typical    |   112.18 ns |  0.39 | 0.0123 |     232 B |        0.23 |
|                              |            |             |       |        |           |             |
| **EncodeTraceIdHeaderValue**     | **worst_case** | **2,051.68 ns** |  **1.00** | **0.3090** |    **5872 B** |        **1.00** |
| EncodeTraceIdHeaderValue_new | worst_case |   143.98 ns |  0.07 | 0.0267 |     504 B |        0.09 |

clean value: `"Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Sampled=1"`
typical value: `Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Sampled=1;Lineage=a87bd80c:1 <tag>`
worst case value: `new string ('\x01', 80)`

## Motivation and Context
Reduce GC pressure and improve performance

## Testing
Covered by existing tests in `RecursionDetectionTests`

### Dry-runs

- **DotNet Dry-run ID:**
  - [ ] Pending
  - [ ] Completed successfully
  - [ ] Failed
- **PowerShell Dry-run ID:**
  - [ ] Pending
  - [ ] Completed successfully
  - [ ] Failed

## Breaking Changes Assessment

1. Identify all breaking changes including the following details:
    * What functionality was changed?
    * How will this impact customers?
    * Why does this need to be a breaking change and what are the most notable non-breaking alternatives?
    * Are best practices being followed?
    * How have you tested this breaking change?
2. Has a senior/+ engineer been assigned to review this PR?

No breaking changes

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

**Comment by peterrsongg** [comment]
Wrote a fuzzer to test the old and new implementation. All edge cases passed 

```
  1. Edge cases — empty strings, control chars, multi-byte UTF-8 (CJK, emoji, accented), null bytes, all 256 chars, long
  strings
  2. Random ASCII — 500K random strings with chars 0–127
  3. Random Unicode — 500K random strings mixing ASCII, 2-byte, and 3-byte UTF-8 characters
```

The performance is better as well.

Kicking off an internal dry run in our build system and will approve when it passes. @JanHyka Can you take a look at the copilot comment though?

File: sdk/src/Core/Amazon.Util/AWSSDKUtils.cs [hunk 1]
```diff
@@ -1255,23 +1267,36 @@ internal static string UrlEncodeSlash(string data)
...
+                        encoded.Append(upperHex[b & 0xF]);
+                    }
                 }
             }
-
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(utf8Bytes);
+            }
             return encoded.ToString();
```
[the remark this answers is not in the document]
Comment (JanHyka) [reply sdk/src/Core/Amazon.Util/AWSSDKUtils.cs]: @peterrsongg sure, just need to wait till I get back to PC, afk for a week.

**Review by dscpinheiro — APPROVED** [review]

**Review by peterrsongg — APPROVED** [review]
